### PR TITLE
Add Skyline support for DDA search of DIA files via DiaUmpire

### PR DESCRIPTION
**Pull request by chambm**
* added ConverterSettingsControl for DiaUmpire
- replaced "Perform DDA Search" radio button with "Input file type" combobox so user can specify whether input files are DIA or DDA (or search results)
- fixed bug reading isolation window offsets from new WIFF2 SDK; we were using the absolute isolation m/z lower and upper bound instead of the lower and upper offset (a value relative to the isolation target m/z)
* fixed "Handle not initialized" bug from WIFF2 destructor when terminating process forcefully (like SeeMS does), which apparently may run the static destructors before non-statics
* updated KeyValueGridDlg to take a parent window and use checkboxes for boolean variables
* switched DiaUmpire to a one-spill-file-per-DIA-window model, where spill files are stored in TEMP; scan metadata is merged/sorted after all windows are processed, but scan data is read on the fly when requested by SpectrumList_DiaUmpire::spectrum()
* moved TemporaryFile from IDPicker::Merger to pwiz::util (Filesystem.cpp/hpp)
* fixed peak picker filter not providing correct centroid/profile metadata for detail levels less than full data
* changed TemporaryDirectory so the default "~SK" prefix can be changed; I made the MSAmanda temp directories all share a common directory under TEMP/~SK_MSAmanda
* updated non-DiaUmpire audit logs
* fixed DiaUmpire::Config::maxThreads not being initialized by parameterless ctor
* moved DiaUmpireTest to SpectrumList_DiaUmpireTest

### CODE REVIEW

**Comment by chambm** [comment]
@brendanx67 @nickshulman I think a code review is in order before merge to master.

File: pwiz_tools/Skyline/Model/AbstractDdaSearchEngine.cs
Comment (nickshulman) [inline pwiz_tools/Skyline/Model/AbstractDdaSearchEngine.cs:45]: I am not an expert on audit log objects, but I was under the impression that if a class had a self-describing ToString method, it would not need to implement IAuditLogObject.

File: pwiz_tools/Skyline/Model/DdaSearch/MSAmandaSearchWrapper.cs
Comment (nickshulman) [inline pwiz_tools/Skyline/Model/DdaSearch/MSAmandaSearchWrapper.cs:219]: This method should really be taking a CancellationToken instead of a CancellationTokenSource (since it is only interesting in asking whether the job has been cancelled and does not need the ability to request cancellation).

That would require changing MSAmandaSearch so that it also took a CancellationToken instead of CancellationTokenSource.

File: pwiz_tools/Skyline/SettingsUI/FullScanSettingsControl.cs
Comment (nickshulman) [inline pwiz_tools/Skyline/SettingsUI/FullScanSettingsControl.cs:904]: Usually for something like this we would replace the control on the form with a new one that we construct. Is there a reason you would not want to do this here?

**Review by nickshulman — COMMENTED** [review]
Looks pretty good.
You might want to take everything in "DiaUmpire.cs" in ProteoWizardWrapper and spread it out across multiple source files in its own folder.
The source file "MsDataFileImpl.cs" is horribly laid out with a lot of things as inner classes as if everyone who added to this class though everything needed to be in the same source file.

File: pwiz_tools/Skyline/Model/DdaSearch/MSAmandaSearchWrapper.cs
Comment (chambm) [reply pwiz_tools/Skyline/Model/DdaSearch/MSAmandaSearchWrapper.cs:219]: I agree but since this requires a change in the MSAmanda DLLs and it's not specifically relevant to DIA-Umpire, let's leave that to a separate change set.

File: pwiz_tools/Skyline/SettingsUI/FullScanSettingsControl.cs
Comment (chambm) [reply pwiz_tools/Skyline/SettingsUI/FullScanSettingsControl.cs:904]: I'm not actually sure this is needed after Kaipo's earlier change with MakeFullScanSettingsControl(). I'll see if it works without it. But I had been trying to call Iniitialize() instead of this and it was not working like I expected (forgot the exact problem though).

File: pwiz_tools/Shared/Common/SystemUtil/FormUtil.cs
Comment (chambm) [reply pwiz_tools/Shared/Common/SystemUtil/FormUtil.cs:67]: Do you mean use the new extension function I added below this one? Because the framework doesn't have a Point function that offsets a point and returns the new point. But yes I agree this function could be briefer by piggybacking off the below one.

**Comment by chambm** [comment]
Also can a few other devs please try running the TestTtofDiaUmpireFullTutorial test? I have noticed some very sporadic crashes but haven't been able to pinpoint the cause or repro conditions.

File: pwiz_tools/Shared/Common/SystemUtil/FormUtil.cs
Comment (nickshulman) [reply pwiz_tools/Shared/Common/SystemUtil/FormUtil.cs:67]: Oops! You're right. I was the one who was confused.

**Comment by chambm** [comment]
> You might want to take everything in "DiaUmpire.cs" in ProteoWizardWrapper and spread it out across multiple source files in its own folder.
Are you really referring to `DiaUmpire.cs`? Because the CS wrapper seems pretty brief to me. The `Config` class is a nested class so to move that out it would either have to be a partial class or un-nested, right?

> The source file "MsDataFileImpl.cs" is horribly laid out with a lot of things as inner classes as if everyone who added to this class though everything needed to be in the same source file.
While I pretty much agree, to be fair the vast majority of that file is consumed by MsDataFileImpl class itself. So separating the helper classes into separate files would only help so much.

File: pwiz_tools/Skyline/FileUI/PeptideSearch/DDASearchControl.cs
Comment (nickshulman) [inline pwiz_tools/Skyline/FileUI/PeptideSearch/DDASearchControl.cs:84]: You should use "MessageDlg.ShowWithException" here instead.

Program.ReportException is only for things that are actual bugs where we want the user to submit a bug report to the exception web site.

**Review by nickshulman — COMMENTED** [review]
Looks really good. Just a couple of comments.

We might want to change this to radio buttons with tooltips instead of having the help text in the combo box options.

image

**Comment by chambm** [comment]
> Looks really good. Just a couple of comments.
> 
> We might want to change this to radio buttons with tooltips instead of having the help text in the combo box options.
> 
> image

What do you think about Nick's suggestion @brendanx67  ?

File: pwiz_tools/Skyline/FileUI/PeptideSearch/DDASearchControl.cs
Comment (chambm) [reply pwiz_tools/Skyline/FileUI/PeptideSearch/DDASearchControl.cs:84]: So `ShowWithException` doesn't allow the user to report the error with a button press? DDA support is still pretty new: I'm thinking that for the near future we should allow users to report these exceptions easily. Perhaps we can just add catches for the common non-bug errors that get reported here, and handle those with `ShowWithException`, but let unexpected errors fall through to `ReportException`?

**Comment by chambm** [comment]
> You might want to take everything in "DiaUmpire.cs" in ProteoWizardWrapper and spread it out across multiple source files in its own folder.

Do you recommend moving the nested Config class to a separate DiaUmpireConfig file? I agree the IterationListenerToMonitor nested class should move to its own file; it's not specific to DiaUmpire at all.

File: pwiz_tools/Skyline/FileUI/PeptideSearch/DDASearchControl.cs
Comment (nickshulman) [reply pwiz_tools/Skyline/FileUI/PeptideSearch/DDASearchControl.cs:84]: You certainly could show a message box with multiple buttons on it, telling them to push one of the buttons if they think the error was a bug (and pushing that button would call Program.ReportException to bring up the ReportErrorDlg). There is no place where we currently do that in the code, but it would be a reasonable thing to do.

You would not want the user to see the ReportErrorDlg without that other message box first. The information on the ReportErrorDlg is presented in a way that is very hard for users to read, so even when the error message is straightforward (e.g. "The file already exists"), users will typically not be able to figure out what went wrong.

**Comment by chambm** [comment]
@brendanx67 @nickshulman Should we make this an integration branch for a few days before merging? Do you want to review yourself, @brendanx67 ?

File: pwiz_tools/Skyline/FileUI/PeptideSearch/ConverterSettingsControl.cs
Comment (chambm) [inline pwiz_tools/Skyline/FileUI/PeptideSearch/ConverterSettingsControl.cs:272]: Doing constructor overload resolution at runtime with a generic parameter required this dynamic voodoo, which required adding the Microsoft.CSharp reference. Any problem with that @nickshulman ?

**Comment by nickshulman** [comment]
There's nothing wrong with doing it that way.
I can't think of any reason that we would need this code to target an
earlier version of the C# language.

In general, we probably should try to make use of as many language features
as possible so that we learn more.
It's rather unnecessary in this case since the value eventually gets
converted to an object. (AbstractDdaSearchEngine.Setting could have a
constructor that takes an object).
-- Nick

**Comment by chambm** [comment]
> (AbstractDdaSearchEngine.Setting could have a constructor that takes an object). 

The typed constructors ensure that the min/max values are the same type as the default/current value. That could be done at runtime with an object/object/object constructor of course, but with overloads you get compile-time checking for most usages.

File: pwiz_tools/Skyline/TestPerf/DiaUmpireTutorialTest.cs
Comment (nickshulman) [inline pwiz_tools/Skyline/TestPerf/DiaUmpireTutorialTest.cs:334]: This assert here fails if you check "Record Audit Logs" in SkylineTester (and the test will not get run in subsequent languages).

If you want to keep this assert here, I think you need to change "IsRecordMode" so that it does something other than return "IsRecordAuditLogForTutorials".

File: pwiz_tools/Skyline/TestPerf/DiaUmpireTutorialTest.cs
Comment (nickshulman) [inline pwiz_tools/Skyline/TestPerf/DiaUmpireTutorialTest.cs:635]: You should use "WaitForOpenForm" here instead of "ShowDialog".

"ShowDialog" will fail if the dialog happens to already be showing.

**Review by nickshulman — APPROVED** [review]
Looks good to me.

**Comment by chambm** [comment]
Ah crap. Forgot to squash. Let me see if I can fix that.

**Comment by chambm** [comment]
@brendanx67  @nickshulman I force pushed to master to squash all those commits. Hopefully I didn't f it up but of course I'll fix it if I did. I noticed a new option in the branch protection options: `Require linear history` which would prevent merging PRs with merge commits. Should we enable that? It's merge commits that cause the wonky lines in the TortoiseGit log.

**Pull request merged**
